src/GafferImage/Resample.cpp

```diff
@@ -0,0 +1,569 @@
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include <iostream>
+
+#include "OpenImageIO/fmath.h"
+#include "OpenImageIO/filter.h"
+
+#include "Gaffer/Context.h"
+#include "Gaffer/StringPlug.h"
+
+#include "GafferImage/Resample.h"
+#include "GafferImage/Sampler.h"
+
+using namespace Imath;
+using namespace IECore;
+using namespace Gaffer;
+using namespace GafferImage;
+
+//////////////////////////////////////////////////////////////////////////
+// Utilities
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+// Used as a bitmask to say which filter pass(es) we're computing.
+enum Passes
+{
+	Horizontal = 1,
+	Vertical = 2,
+	Both = Horizontal | Vertical
+};
+
+unsigned requiredPasses( const Resample *resample, const ImagePlug *image, const OIIO::Filter2D *filter )
+{
+	int debug = resample->debugPlug()->getValue();
+	if( debug == Resample::HorizontalPass )
+	{
+		return Horizontal;
+	}
+	else if( debug == Resample::SinglePass )
+	{
+		return Horizontal | Vertical;
+	}
+
+	if( image == image->parent<ImageNode>()->outPlug() )
+	{
+		return filter->separable() ? Vertical : Both;
+	}
+	return Horizontal;
+}
+
+// Rounds min down, and max up, while converting from float to int.
+Box2i box2fToBox2i( const Box2f &b )
+{
+	return Box2i(
+		V2i( floor( b.min.x ), floor( b.min.y ) ),
+		V2i( ceil( b.max.x ), ceil( b.max.y ) )
+	);
+}
+
+// Calculates the scale and offset needed to convert from output
+// coordinates to input coordinates.
+void ratioAndOffset( const Box2f &dstDataWindow, const Box2i &srcDataWindow, V2f &ratio, V2f &offset )
+{
+	/// \todo Adjust for #1438.
+	const V2f dstSize = dstDataWindow.size() + V2i( 1 );
+	const V2f srcSize = srcDataWindow.size() + V2i( 1 );
+
+	ratio = dstSize / srcSize;
+	offset = srcDataWindow.min - dstDataWindow.min / ratio;
+}
+
+// Returns the input region that will need to be sampled when
+// generating a given output tile.
+Box2i inputRegion( const V2i &tileOrigin, unsigned passes, const V2f &ratio, const V2f &offset, const OIIO::Filter2D *filter )
+{
+	Box2f outputRegion( V2f( tileOrigin ), tileOrigin + V2f( ImagePlug::tileSize() ) );
+	V2f filterRadius( filter->width() / 2.0f, filter->height() / 2.0f );
+
+	Box2f result = outputRegion;
+	if( passes & Horizontal )
+	{
+		result.min.x = result.min.x / ratio.x + offset.x - filterRadius.x;
+		result.max.x = result.max.x / ratio.x + offset.x + filterRadius.x;
+	}
+	if( passes & Vertical )
+	{
+		result.min.y = result.min.y / ratio.y + offset.y - filterRadius.y;
+		result.max.y = result.max.y / ratio.y + offset.y + filterRadius.y;
+	}
+
+	return box2fToBox2i( result );
+}
+
+typedef boost::shared_ptr<OIIO::Filter2D> Filter2DPtr;
+Filter2DPtr createFilter( const std::string &name, const V2f &filterWidth, const V2f &ratio )
```

Comment:
Oh, and these issues only seem to be when upscaling - they all seem good when downscaling.